File: restty-app/frontend/src/main/frontend/src/app/services/test-case.service.ts
import { Stats } from '../model/stats';
import { TestCase } from '../model/test-case';
import { ProjectService } from './project.service';
import { HttpParams } from '@angular/common/http';
import { HttpHeaders } from '@angular/common/http';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

const httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
};

@Injectable()
export class TestCaseService {

  public static get TEST_CASES_PATH(): string { return '/test-cases'; }
  public static get TEST_CASE_PATH(): string { return '/api/test-cases'; }
  public static get TEST_CASE_VALIDATION_PATH(): string { return TestCaseService.TEST_CASES_PATH + '/validate'; }

  public static get TEST_CASE_RUN_PATH(): string { return '/run'; }

  constructor(private http: HttpClient) {}

  /**
   * Finds all test cases for given project.
   *
   * @param projectId ID of project to search by
   * @return list of test cases
   */
  findAll(projectId: number) {
    return this.http.get<TestCase[]>(ProjectService.PROJECTS_PATH + `/${projectId}` + TestCaseService.TEST_CASES_PATH);
  }

  /**
   * Finds test case by given ID.
   *
   * @param testCaseId ID of test case to search by
   * @return TestCase
   */
  findById(testCaseId: number) {
    return this.http.get<TestCase>(TestCaseService.TEST_CASE_PATH + `/${testCaseId}`);
  }

  /**
   * Finds test case by given name.
   *
   * @param projectId ID of project to search by
   * @param name Name to search test case by
   * @return TestCase
   */
  findByName(projectId: number, name: string) {
    let params = new HttpParams();
    params = params.append('name', name);
    return this.http.get(ProjectService.PROJECTS_PATH + `/${projectId}` + TestCaseService.TEST_CASE_VALIDATION_PATH, {params: params});
  }

  /**
   * Finds test case by given ID and runs it.
   *
   * @param testCaseId ID of test case to run.
   * @return response status
   */
  run(testCaseId: number) {
    return this.http.post(TestCaseService.TEST_CASE_PATH + `/${testCaseId}` + TestCaseService.TEST_CASE_RUN_PATH, null);
  }

  /**
   * Finds test cases with given IDs and runs them.
   *
   * @param testCaseIds IDs of test cases to run
   * @return response status
   */
  runAll(testCaseIds: number[]) {
    return this.http.post('/api' + TestCaseService.TEST_CASES_PATH + TestCaseService.TEST_CASE_RUN_PATH, testCaseIds, httpOptions);
  }

  /**
   * Creates new test case from the information in the input parameter.
   *
   * @param projectId ID of project to create test case for
   * @param testCase Information about new TestCase
   * @return TestCase
   */
  createTestCase(projectId: number, testCase: TestCase) {
    return this.http.post<TestCase>(ProjectService.PROJECTS_PATH + `/${projectId}`
       + TestCaseService.TEST_CASES_PATH, JSON.stringify(testCase), httpOptions);
  }

  /**
   * Deletes test cases with given IDs from the project.
   *
   * @param projectId ID of the project
   * @param testCasesIds IDs of the test cases to delete
   */
  deleteTestCases(projectId: number, testCasesIds: number[]) {
    return this.http.request('delete', ProjectService.PROJECTS_PATH + `/${projectId}` + TestCaseService.TEST_CASES_PATH, {
      headers: httpOptions.headers,
      body: testCasesIds
    });
  }

}

File: restty-app/backend/src/main/java/cz/restty/app/service/TestCaseService.java
package cz.restty.app.service;

import java.util.List;

import cz.restty.app.entities.Project;
import cz.restty.app.entities.TestCase;
import cz.restty.app.rest.dto.TestCaseDto;

/**
 * Service layer for manipulation with {@link TestCase} entities.
 * 
 * @author Ondrej Krpec
 *
 */
public interface TestCaseService {

    /**
     * Runs given testCase against the test server.
     * 
     * @param testCase
     *            TestCase to run.
     * @return True if run was successful, false otherwise.
     */
    boolean run(TestCase testCase);


    /**
     * Creates {@link TestCase} in given project.
     * 
     * @param project
     *            {@link Project} to create test case for.
     * @param testCaseDto
     *            DTO that contains information about new test case.
     * @return new {@link TestCaseDto}
     */
    TestCaseDto createTestCase(Project project, TestCaseDto testCaseDto);

    /**
     * Deletes all test cases for given project.
     * 
     * @param project
     *            {@link Project} to delete test cases for.
     */
    void deleteAllByProject(Project project);

    /**
     * Deletes test cases with given IDs that belong to the project.
     * 
     * @param project
     *            {@link Project}
     * @param testCasesIds
     *            IDs of test cases to delete
     */
    void deleteTestCases(Project project, List<Long> testCasesIds);

}
